### PR TITLE
iscsi:new function iscsi_get_device_by_lun

### PR DESCRIPTION
**Pull request by aliang123**
function:iscsi_get_device_by_lun() to get
logged-in iscsi device by lun.

Signed-off-by: Aihua Liang <aliang@redhat.com>
id:2100698

### CODE REVIEW

**Comment by aliang123** [comment]
@zhencliu @YongxueHong, please help to review, thanks.

File: virttest/iscsi.py [hunk 1]
```diff
@@ -330,14 +330,18 @@ def get_device_name(self):
...
             output = process.run(cmd).stdout_text
             targets = output.split('Target: ')[1:]
             for target in targets:
                 if self.target in target:
-                    device_name = re.findall(pattern, target, re.S)
+                    self.devices = dict(re.findall(pattern, target, re.S))
+                    device_name.append(self.devices["0"])
+                    for lun, device in self.devices.items():
+                        dname = "/dev/%s" % device
+                        self.devices[lun] = dname
```
Comment (zhencliu) [inline virttest/iscsi.py:344]: hi Aihua, I just noticed that we need to call get_device_name first, then we can get self.devices, i.e. self.devices depends on get_device_name function, that might not be a good idea. So what about introducing a new interface, like get_devices, which returns the {lun: dev}. In this way, we don't need self.devices, every time we use it, we call the function to get it. For the devices could change(think about that we add a new lun in server, we have to refresh devices)

Comment (aliang123) [reply virttest/iscsi.py]: updated, @zhencliu, please help to review again, thanks.

File: virttest/iscsi.py [hunk 2]
```diff
@@ -347,6 +347,26 @@ def get_device_name(self):
             LOG.error("Session is not logged in yet.")
         return device_name
 
+    def get_devices(self):
+        """
+        Get devices from the target name.
+        """
+        cmd = "iscsiadm -m session -P 3"
+        pattern = r"Lun:\s(\d+).*?scsi\sdisk\s(\w+)\s+State:\srunning"
```
Comment (zhencliu) [inline virttest/iscsi.py:355]: the pattern doesn't work well for the following:
Attached SCSI devices:
************************
Host Number: 17 State: running
scsi17 Channel 00 Id 0 Lun: 0
Attached scsi disk sdf          State: running
scsi17 Channel 00 Id 0 Lun: 1
Attached scsi disk sdi          State: ing         >>>>>>not running here
scsi17 Channel 00 Id 0 Lun: 2
Attached scsi disk sdh          State: running
scsi17 Channel 00 Id 0 Lun: 3
Attached scsi disk sdg          State: running

I cannot find an easy way to handle this, just fyi.
pattern = r"Lun:\s(\d+).*?scsi\sdisk\s(\w+)\s+State:\s(\w+)"
devices = {t[0]:t[1] for t in re.findall(pattern, c, re.S) if t and t[2] == 'running'}

File: virttest/iscsi.py [hunk 3]
```diff
@@ -347,6 +347,26 @@ def get_device_name(self):
...
+        devices = None
+        if self.logged_in():
+            output = process.run(cmd).stdout_text
+            targets = output.split('Target: ')[1:]
+            for target in targets:
+                if self.target in target:
+                    devices = dict(re.findall(pattern, target, re.S))
+                    for lun, device in devices.items():
+                        dname = "/dev/%s" % device
+                        devices[lun] = dname
```
Comment (zhencliu) [inline virttest/iscsi.py:365]: do we want to break out the for target in targets loop inside if?

**Comment by aliang123** [comment]
Close this patch for other solution has been adopted.

**Comment by aliang123** [comment]
Close this patch for other solution has been adopted.

**Pull request closed without merging**